ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2962,286 @@ def __init__(self, *args, **kwargs):
+            props.append("{}={!r}".format(prop._code_name, prop))
+        return "{}<{}>".format(cls.__name__, ", ".join(props))
+
+
+class Model(metaclass=MetaModel):
+    """A class describing Cloud Datastore entities.
+
+    Model instances are usually called entities. All model classes
+    inheriting from :class:`Model` automatically have :class:`MetaModel` as
+    their metaclass, so that the properties are fixed up properly after the
+    class is defined.
+
+    Because of this, you cannot use the same :class:`Property` object to
+    describe multiple properties -- you must create separate :class:`Property`
+    objects for each property. For example, this does not work:
+
+    .. code-block:: python
+
+        reuse_prop = ndb.StringProperty()
+
+        class Wrong(ndb.Model):
+            first = reuse_prop
+            second = reuse_prop
+
+    instead each class attribute needs to be distinct:
+
+    .. code-block:: python
+
+        class NotWrong(ndb.Model):
+            first = ndb.StringProperty()
+            second = ndb.StringProperty()
+
+    The "kind" for a given :class:`Model` subclass is normally equal to the
+    class name (exclusive of the module name or any other parent scope). To
+    override the kind, define :meth:`_get_kind`, as follows:
+
+    .. code-block:: python
+
+        class MyModel(ndb.Model):
+            @classmethod
+            def _get_kind(cls):
+                return "AnotherKind"
+
+    A newly constructed entity will not be persisted to Cloud Datastore without
+    an explicit call to :meth:`put`.
+
+    .. note::
+
+        You cannot define a property named ``key``; the ``.key`` attribute
+        always refers to the entity's ``key``. But you can define properties
+        named ``id`` or ``parent``. Values for the latter cannot be passed
+        through the constructor, but can be assigned to entity attributes
+        after the entity has been created.
+
+    .. automethod:: _get_kind
+
+    Args:
+        key (Key): Datastore key for this entity (kind must match this model).
+            If ``key`` is used, ``id`` and ``parent`` must be unset or
+            :data:`None`.
+        id (str): Key ID for this model. If ``id`` is used, ``key`` must be
+            :data:`None`.
+        parent (Key): The parent model or :data:`None` for a top-level model.
+            If ``parent`` is used, ``key`` must be :data:`None`.
+        namespace (str): Namespace for the entity key.
+        app (str): Application ID for the entity key.
+        kwargs (Dict[str, Any]): Additional keyword arguments. These should map
+            to properties of this model.
+
+    Raises:
+        .BadArgumentError: If the constructor is called with ``key`` and one
+            of ``id``, ``app``, ``namespace`` or ``parent`` specified.
+    """
+
+    # Class variables updated by _fix_up_properties()
+    _properties = None
+    _has_repeated = False
+    _kind_map = {}  # Dict mapping {kind: Model subclass}
```

Comment:
IMO this kind map should be gardened on `MetaModel` and reads and writes to it should be threadsafe. But it's probably not worth the yak-shaving / bikeshedding since it will (**should**) only be modified at code import time.

Reply:
Yeah, I'm fine with it as-is.